library/core/src/num/int_macros.rs

```diff
@@ -1231,6 +1234,36 @@ macro_rules! int_impl {
             (a as Self, b)
         }
 
+        /// Calculates `self` + `rhs` + `carry`.
+        ///
+        /// Performs "ternary addition" which takes in an extra bit to add, and may return an
+        /// additional bit of overflow. This allows for chaining together multiple additions
+        /// to create "big integers" which represent larger values.
+        ///
+        /// # Examples
+        ///
+        /// Basic usage
+        ///
+        /// ```
+        /// #![feature(carrying)]
+        #[doc = concat!("assert_eq!(5", stringify!($SelfT), ".carrying_add(2, false), (7, false));")]
+        #[doc = concat!("assert_eq!(5", stringify!($SelfT), ".carrying_add(2, true), (8, false));")]
+        #[doc = concat!("assert_eq!(", stringify!($SelfT), "::MAX.carrying_add(1, false), (0, true));")]
+        #[doc = concat!("assert_eq!(", stringify!($SelfT), "::MAX.carrying_add(1, true), (1, true));")]
+        /// ```
+        #[unstable(feature = "carrying", issue = "none")]
+        #[rustc_const_unstable(feature = "const_carrying", issue = "none")]
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline(always)]
+        pub const fn carrying_add(self, rhs: Self, carry: bool) -> (Self, bool) {
+            // note: longer-term this should be done via an intrinsic, but this has been shown
+            //   to generate optimal code for now, and LLVM doesn't have an equivalent intrinsic
+            let (a, b) = self.overflowing_add(rhs);
+            let (c, d) = a.overflowing_add(carry as $SelfT);
+            (c, b | d)
+        }
```

Comment:
To make my point clearer:

If by carry you mean overflow, then both should return `false`, but then it is not carry and cannot be used easily for a wider addition. So that would make this function pretty useless.

If by carry you actually mean carry, then both should return `true`. The implementation would be something like:
```rust
let (u, c) = (self as $UnsignedT).carrying_add(rhs as $UnsignedT, carry);
(u as $SelfT, c)
```

(To me the current `carrying_add` looks completely fine for unsigned.)

Reply:
(I did make that latter change by the way, but will leave you to mark this as resolved in case I misread what you said)

Reply:
I see this is now resolved (including borrowing subtraction) but I don't think I can mark it as resolved myself.